master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -624,12 +625,34 @@ private[celeborn] class Master(
     context.reply(OneWayMessageResponse)
   }
 
+  private def computeUserResourceConsumption(userIdentifier: UserIdentifier)
+      : ResourceConsumption = {
+    statusSystem.workers.asScala.flatMap { workerInfo =>
+      workerInfo.userResourceConsumption.asScala.get(userIdentifier)
+    }.foldRight(ResourceConsumption(0, 0, 0, 0))(_ add _)
+  }
+
   private def handleCheckQuota(
       userIdentifier: UserIdentifier,
       context: RpcCallContext): Unit = {
-    val userResourceConsumption = statusSystem.workers.asScala.flatMap { workerInfo =>
-      workerInfo.userResourceConsumption.asScala.get(userIdentifier)
-    }.foldRight(ResourceConsumption(0, 0, 0, 0))(_ add _)
+    resourceConsumptionSource.addGauge(
+      "diskFIleCount",
+      _ => computeUserResourceConsumption(userIdentifier).diskFileCount,
+      userIdentifier.toMap)
+    resourceConsumptionSource.addGauge(
+      "diskBytesWritten",
+      _ => computeUserResourceConsumption(userIdentifier).diskBytesWritten,
+      userIdentifier.toMap)
+    resourceConsumptionSource.addGauge(
+      "hdfsFileCount",
+      _ => computeUserResourceConsumption(userIdentifier).hdfsFileCount,
+      userIdentifier.toMap)
+    resourceConsumptionSource.addGauge(
+      "hdfsBytesWritten",
+      _ => computeUserResourceConsumption(userIdentifier).hdfsBytesWritten,
+      userIdentifier.toMap)
```

Comment:
Could we run computeUserResourceConsumption() only once when each time collecting gauge metrics? I am worry that too many repeated calculations will cause an additional burden to Master, especially when there are lots of user in cluster.

Reply:
> 

Agree. We can write as follows
```
    val userResourceConsumption = computeUserResourceConsumption(userIdentifier)
    resourceConsumptionSource.addGauge(
      "diskFIleCount",
      _ => userResourceConsumption.diskFileCount,
      userIdentifier.toMap)
    resourceConsumptionSource.addGauge(
      "diskBytesWritten",
      _ => userResourceConsumption.diskBytesWritten,
      userIdentifier.toMap)
    resourceConsumptionSource.addGauge(
      "hdfsFileCount",
      _ => userResourceConsumption.hdfsFileCount,
      userIdentifier.toMap)
    resourceConsumptionSource.addGauge(
      "hdfsBytesWritten",
      _ => userResourceConsumption.hdfsBytesWritten,
      userIdentifier.toMap)

    val quota = quotaManager.getQuota(userIdentifier)
    val (isAvailable, reason) =
      quota.checkQuotaSpaceAvailable(userIdentifier, userResourceConsumption)
    context.reply(CheckQuotaResponse(isAvailable, reason))
```

Reply:
In fact, I think we can cache computeUserResourceConsumption for say 30s period in master to further release the burden. Master periodically compute the consumption, handleCheckQuota merely reads it.

Reply:
> In fact, I think we can cache computeUserResourceConsumption for say 15s period in master to further release the burden, with the tradeoff of a little out-of-date. Master periodically compute the consumption, handleCheckQuota merely reads it.

Done, how about current?